include/aspect/utilities.h

```diff
@@ -1069,6 +1071,51 @@ namespace aspect
     template <int dim>
     SymmetricTensor<2,dim> nth_basis_for_symmetric_tensors (const unsigned int k);
 
+    /**
+     * This function returns if a given plugin (e.g. a material model returned
+     * from SimulatorAccess::get_material_model() ) matches a certain plugin
+     * type (e.g. MaterialModel::Simple). This check is needed, because often
+     * it is only possible to get a reference to an Interface, not the actual
+     * plugin type, but the actual plugin type might be important. For example
+     * a radial gravity model might only be implemented for spherical geometry
+     * models, and would want to check if the current geometry is in fact a
+     * spherical shell.
+     */
+    template <typename TestType, typename PluginType>
+    inline
+    bool
+    plugin_type_matches (const PluginType &object)
+    {
+      return (dynamic_cast<const TestType *> (&object) != 0);
+    }
+
+    /**
+     * This function converts a reference to a type (in particular a reference
+     * to an interface class) into a reference to a different type (in
+     * particular a plugin class). This allows accessing members of the plugin
+     * that are not specified in the interface class. Note that you should
+     * first check if the plugin type is actually convertible by calling
+     * plugin_matches_type() before calling this function. If the plugin is
+     * not convertible this function throws an exception.
+     */
+    template <typename TestType, typename PluginType>
+    inline
+    TestType &
+    get_plugin_as_type (PluginType &object)
+    {
+      if (!plugin_type_matches<TestType>(object))
+        AssertThrow(false,
+                    ExcMessage("You have requested to convert a plugin of type <"
+                               + boost::core::demangle(typeid(PluginType).name())
+                               + "> into type <"
+                               + boost::core::demangle(typeid(TestType).name()) +
+                               "> in which it can not be converted."));
```

Comment:
this reads awkwardly. say "...into type <...>. But this cast cannot be performed."